lib/Sema/CSSimplify.cpp

```diff
@@ -3626,33 +3626,36 @@ ConstraintSystem::matchExistentialTypes(Type type1, Type type2,
 
           if (last.is<LocatorPathElt::ApplyArgToParam>()) {
             auto proto = protoDecl->getDeclaredInterfaceType();
-            ConstraintFix *fix;
             // Impact is 2 here because there are two failures
             // 1 - missing conformance and 2 - incorrect argument type.
             //
             // This would make sure that arguments with incorrect
             // conformances are not prioritized over general argument
             // mismatches.
-            unsigned impact;
             if (type1->isOptional()) {
               auto unwrappedType = type1->lookThroughAllOptionalTypes();
-              auto matchTypeResult =
-                  matchTypes(unwrappedType, type2, ConstraintKind::Conversion,
-                             subflags, locator);
-              if (matchTypeResult.isSuccess()) {
-                // Impact here is a 1 because there is only one failure before
-                // we need to force type1 to be
-                impact = 1;
-                fix = ForceOptional::create(*this, type1, proto,
-                                            getConstraintLocator(locator));
+              auto result = simplifyConformsToConstraint(
+                  unwrappedType, type2, ConstraintKind::ConformsTo,
+                  getConstraintLocator(locator),
+                  TypeMatchFlags::TMF_ApplyingFix);
+              if (result == SolutionKind::Solved) {
+                auto matchTypeResult =
+                    matchTypes(unwrappedType, type2, ConstraintKind::Conversion,
+                               subflags, locator);
+                if (matchTypeResult.isSuccess()) {
+                  // Impact here is a 1 because there is only one failure before
+                  // we need to force type1 to be
+                  auto fix = ForceOptional::create(
+                      *this, type1, proto, getConstraintLocator(locator));
+                  if (recordFix(fix, /*impact=*/1))
```

Comment:
Nit: One is already the default for `recordFix` so we don't need to pass it. Also we probably don't need the comment above as well.

Reply:
Besides that result could be three different things - solved, failed, unsolved. It's possible that the type here would be `$T?` where `$T` is a type variable, so we actually need to check that and delay a decision until optional is fully resolved.

Reply:
@xedin Is that possible in this code path? I mean, this is at the error path for the first call to [`simplifyConformsToConstraint`](https://github.com/apple/swift/blob/a1fe5001f09256a2241a4f0f324388ca5014b576/lib/Sema/CSSimplify.cpp#L3602) so if existential is not resolved yet, it would never reach diagnostic logic anyways(first call would return unsolved). So I think we don't need to check that here because that would never be the case.

Reply:
I was thinking about the case involving primary associated types where the conformance is conditional on such a type that is not yet resolved...

Reply:
Ah I see, but still not sure how we are going to end up at this code path with unresolved variables. Do you have an example?